File: algorithms/lab00/python/task-2/sol1_sliding_window.py
# SPDX-License-Identifier: BSD-3-Clause

import bisect
from collections import deque

class Solution:
    # Solutie: Folosim un multiset pentru a mentine cele mai mari k elemente.
    # Complexitate: O(nlogk)
    @staticmethod
    def max_sliding_window(nums: list[int], k: int) -> list[int]:
        # sol stocheaza al k-lea cel mai mare element din fiecare fereastra.
        sol: list[int] = []
        # window stocheaza elementele din fereastra curenta.
        window: deque[int] = deque()
        # heap este un multiset care stocheaza cele mai mari k elemente.
        heap: list[int] = []

        for x in nums:
            # Insereaza primele k elemente in multiset.
            if len(window) < k:
                window.append(x)
                bisect.insort(heap, x)
                continue

            # Insereaza al k-lea cel mai mare element in sol.
            sol.append(heap[-1])

            # Elimina cel mai vechi element din fereastra si din multiset.
            old_x = window.popleft()
            pos = bisect.bisect_left(heap, old_x)
            del heap[pos]

            window.append(x)
            bisect.insort(heap, x)
        # Insereaza al k-lea cel mai mare element din ultima fereastra in sol.
        sol.append(heap[-1])
        return sol

File: algorithms/lab00/rust/task-2/sol1_sliding_window.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::collections::btree_map::Entry;
use std::collections::{BTreeMap, VecDeque};

// Solutie: Folosim un BTreeMap pentru a mentine cele mai mari k elemente din fereastra.
// BTreeMap e ordonat dupa cheie; maximul = last_key_value().
// Complexitate: O(n log k)

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let k = k as usize;
        let mut sol: Vec<i32> = Vec::new();
        let mut window: VecDeque<i32> = VecDeque::new();
        // heap: multiset (element -> count) ordonat dupa cheie; maximul = ultima cheie
        let mut heap: BTreeMap<i32, usize> = BTreeMap::new();

        for x in nums {
            if window.len() < k {
                window.push_back(x);
                *heap.entry(x).or_insert(0) += 1;
                continue;
            }

            // Insereaza maximul din fereastra curenta in sol.
            let max_key = *heap.keys().next_back().unwrap();
            sol.push(max_key);

            // Elimina cel mai vechi element din fereastra si din multiset.
            let old_x = window.pop_front().unwrap();
            if let Entry::Occupied(mut e) = heap.entry(old_x) {
                *e.get_mut() -= 1;
                if *e.get() == 0 {
                    e.remove_entry();
                }
            }

            window.push_back(x);
            *heap.entry(x).or_insert(0) += 1;
        }
        // Ultima fereastra: maximul
        let max_key = *heap.keys().next_back().unwrap();
        sol.push(max_key);

        sol
    }
}
